crates/matrix-sdk-crypto/src/olm/account.rs

```diff
@@ -1511,12 +1509,17 @@ impl Account {
 
             let encryption_info = Self::get_olm_encryption_info(sender_key, sender, &sender_device);
 
-            Ok(DecryptionResult {
-                event,
-                raw_event: Raw::from_json(RawJsonValue::from_string(plaintext)?),
-                sender_key,
-                encryption_info,
-            })
+            let parsed: Value = serde_json::from_str(&plaintext)?;
+
+            let raw_event = json!({
+                "type": event.event_type().to_owned(),
+                "content": parsed["content"].clone(),
+                "sender": event.sender().to_owned(),
+            });
+
+            let raw_event: Raw<AnyToDeviceEvent> = Raw::from_json_string(raw_event.to_string())?;
+
+            Ok(DecryptionResult { event, raw_event, sender_key, encryption_info })
```

Comment:
It doesn't look super efficient. You are deserializing the entire `plaintext` as `Value`, then you are creating another `Value` based on the previous one, then you're serializing it with `to._string`, to finally re-deserializing it here.

I wonder if you could not “simply” replace `Raw::from_json(RawJsonValue::from_string(…))` by `Raw::<AnyToDeviceEvent>::from_json_string(plaintext);`, something like this:

```suggestion
            Ok(DecryptionResult { event, raw_event: Raw::<AnyToDeviceEvent>::from_json_string(plaintext)?, sender_key, encryption_info })
```

Not tested.

Thoughts?